File: sync/src/main/scala/com/fan/mysql/parser/RotateEventParser.java
package com.fan.mysql.parser;


import com.fan.mysql.binlog.BinlogEventParser;
import com.fan.mysql.dbsync.BinlogContext;
import com.fan.mysql.dbsync.LogBuffer;
import com.fan.mysql.event.BinlogEvent;
import com.fan.mysql.event.EventHeader;
import com.fan.mysql.event.impl.RotateEvent;
import com.fan.mysql.util.ByteUtil;

@SuppressWarnings("unused")
public class RotateEventParser implements BinlogEventParser {

    public BinlogEvent parse(LogBuffer buffer, EventHeader header, BinlogContext context) {
        long position = buffer.getLong64();
        int fileLen = buffer.limit() - buffer.position();
        String file = buffer.getFixString(fileLen);
        RotateEvent event = new RotateEvent(header);
        event.setBinlogFile(file);
        event.setBinlogPosition(position);
        context.setBinlogFileName(file);
        return event;
    }

    public static byte[] getBody(String fileName, long position) {
        byte[] body = new byte[fileName.length() + 8];
        ByteUtil.int8store(body, 0, position);
        byte[] b = fileName.getBytes();
        System.arraycopy(b, 0, body, 8, b.length);
        return body;
    }

}

File: async/src/main/scala/com/fan/mysql/async/binlog/parse/RotateEventParser.scala
package com.fan.mysql.async.binlog.parse

import com.fan.mysql.async.binlog.event.impl.RotateEvent
import com.fan.mysql.async.binlog.event.{BinlogEvent, EventHeader}
import com.fan.mysql.async.binlog.{BinlogDumpContext, BinlogEventParser}
import com.fan.mysql.async.util.ChannelWrapper._
import io.netty.buffer.ByteBuf

class RotateEventParser extends BinlogEventParser {

  override def parse(buffer: ByteBuf, header: EventHeader, context: BinlogDumpContext): BinlogEvent = {
    val position = buffer.readLong()
    val fileLen = buffer.writerIndex() - buffer.readerIndex()
    val file = buffer.readFixedASCIString(fileLen)
    val event = new RotateEvent(header)
    event.setBinlogFile(file)
    event.setBinlogPosition(position)
    context.setBinlogFileName(file)
    event
  }
}
